cirq-core/cirq/work/multiprocessing.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import (
+    Any,
+    Union,
+    List,
+    Callable,
+    TypeVar,
+    Iterable,
+    Optional,
+    ContextManager,
+    TYPE_CHECKING,
+)
+
+import concurrent.futures
+import tqdm
+
+if TYPE_CHECKING:
+    import multiprocessing
+
+_OUTPUT_T = TypeVar('_OUTPUT_T')
+
+
+def execute_with_progress_bar(
+    func: Callable[..., _OUTPUT_T],
+    inputs: Iterable[Any],
+    pool: Optional[
+        Union['multiprocessing.pool.Pool', concurrent.futures.ThreadPoolExecutor]
+    ] = None,
+    progress_bar: Callable[..., ContextManager] = tqdm.tqdm,
+    **progress_bar_args,
+) -> List[_OUTPUT_T]:
+    """Execute a function in parallel with progress bar.
+
+    Args:
+        func: The callable to execute, the function takes a single argument.
+        inputs: An iterable of the argument to pass to the function.
+        pool: An optional multiprocessing or threading pool.
+        progress_bar: A callable that creates a progress bar.
+        progress_bar_args: Arguments to pass to the progress bar.
+
+    Returns:
+        An out-of-order list of the results of the function calls.
+    """
+    sequential_inputs = [*inputs]
```

Comment:
Nit - better to be verbose and more readable
```suggestion
    sequential_inputs = list(inputs)
```